cirq/iqm/iqm_device_test.py

```diff
@@ -0,0 +1,131 @@
+import pytest
+import cirq
+from cirq import iqm
+
+
+class TestOperationValidation:
+    adonis = iqm.Adonis()
+    q0 = cirq.GridQubit(0, 1)
+    q1 = cirq.GridQubit(1, 0)
+    q2 = cirq.GridQubit(1, 1)
+    q3 = cirq.GridQubit(1, 2)
+    q4 = cirq.GridQubit(2, 1)
+
+    def test_valid_operations(self):
+        self.adonis.validate_operation(cirq.X(self.q0))
+        self.adonis.validate_operation(cirq.Y(self.q2))
+        self.adonis.validate_operation(cirq.Z(self.q4).with_tags('tag_foo'))
+
+        self.adonis.validate_operation(cirq.YPowGate(exponent=0.25)(self.q0))
+
+        self.adonis.validate_operation(cirq.CZ(self.q1, self.q2))
+
+        self.adonis.validate_operation(
+            cirq.CZPowGate(exponent=0.5)(self.q1, self.q2))
+
+        self.adonis.validate_operation(cirq.measure(self.q0))
+        self.adonis.validate_operation(cirq.measure(self.q1, key='test'))
+
+    def test_invalid_operations(self):
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(cirq.H(self.q0))
+
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(
+                cirq.PauliString([cirq.X(self.q0),
+                                  cirq.Y(self.q0)]))
+
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(cirq.CNOT(self.q1, self.q2))
+
+    def test_qubits_not_on_device(self):
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(cirq.X(cirq.GridQubit(0, 0)))
+
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(
+                cirq.CZ(cirq.GridQubit(2, 0), cirq.GridQubit(2, 1)))
+
+    def test_qubits_not_connected(self):
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(cirq.CZ(self.q0, self.q3))
+
+        with pytest.raises(ValueError):
+            self.adonis.validate_operation(
+                cirq.CZPowGate(exponent=0.11)(self.q1, self.q3))
+
+
+class TestGateDecomposition:
+    adonis = iqm.Adonis()
+    q0 = cirq.GridQubit(0, 1)
+    q1 = cirq.GridQubit(1, 0)
+    q2 = cirq.GridQubit(1, 1)
+    q3 = cirq.GridQubit(1, 2)
+    q4 = cirq.GridQubit(2, 1)
+
+    @staticmethod
+    def is_native(op_or_op_list) -> bool:
+        if iqm.Adonis.is_native_operation(op_or_op_list):
+            return True
+        for op in op_or_op_list:
+            if not iqm.Adonis.is_native_operation(op):
+                return False  # coverage: ignore
```

Comment:
Why are we ignoring coverage here?

Reply:
That line is only executed in case of a failing test, and of course we don't have any failing tests :)

Reply:
One way to fix this is to change it to something like this:

```
    @staticmethod
    def assert_native(op_or_op_list):
        if iqm.Adonis.is_native_operation(op_or_op_list):
            assert True
        for op in op_or_op_list:
            assert iqm.Adonis.is_native_operation(op)
```